odoo/models.py

```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+
+        if granularity and field.type not in ('datetime', 'date'):
+            raise ValueError(f"Granularity set on a no-datetime field: {groupby_spec!r}")
+
+        if not field.store and not field.inherited:  # TODO: should be manage by _inherits_join_calc
+            raise ValueError(f"Groupby on no-store field: {groupby_spec!r}")
+
+        field_expression = self._inherits_join_calc(self._table, fname, query)
+        if field.type == 'datetime' and self.env.context.get('tz') in pytz.all_timezones_set:
+            # TODO: '{self._context.get('tz')}' should be passed as args
+            field_expression = f"timezone('{self.env.context.get('tz')}', timezone('UTC', {field_expression}))"
+
+        if field.type in ('datetime', 'date'):
+            if not granularity:
+                raise ValueError(f"Granularity not set on a date(time) field: {groupby_spec!r}")
+            if granularity not in TIME_GRANULARITY_AGGREGATION:
+                raise ValueError(f"Granularity specification isn't correct: {granularity!r}")
+
+            # TODO: '{granularity}' should be passed as args
+            if granularity == 'week':
+                # first_week_day: 0=Monday, 1=Tuesday, ...
+                first_week_day = int(get_lang(self.env).week_start) - 1
+                days_offset = first_week_day and 7 - first_week_day
+                field_expression = f"(date_trunc('week', {field_expression}::timestamp - INTERVAL '-{days_offset} DAY') + INTERVAL '-{days_offset} DAY')"
+            else:
+                field_expression = f"date_trunc('{granularity}', {field_expression}::timestamp)"
+            if field.type == 'date':
+                field_expression += '::date'
+        elif field.type == 'boolean':
+            field_expression = f"COALESCE({field_expression}, FALSE)"
+
+        return field_expression, [fname]
+
+    @api.model
+    def _read_group_having(self, query, having_domain):
+        fnames_used = []
+        having_params = []
+        stack_op = []
+
+        def pop_next_condition():
+            stack_leaf = stack_op.pop()
+            if isinstance(stack_leaf, str):
+                return stack_leaf
+
+            aggragate_spec, ope, value = stack_leaf
+            having_params.append(value)
+            select_term, select_fnames_used = self._read_group_select(query, aggragate_spec)
+            fnames_used.extend(select_fnames_used)
+            return f'{select_term} {ope} %s'
+
+        for leaf in reversed(having_domain):
+            if leaf == '!':
+                condition_1 = pop_next_condition()
+                stack_op.append(f'(NOT {condition_1})')
+            elif leaf in ('&', '|'):
+                condition_1 = pop_next_condition()
+                condition_2 = pop_next_condition()
+
+                op = 'AND' if leaf == '&' else 'OR'
+                stack_op.append(f'({condition_1} {op} {condition_2})')
+            elif isinstance(leaf, (list, tuple)):
```

Comment:
could be nice to add a test about my `'"="'` trick if you remember

Reply:
Already done in the `test_malformed_params`